Support zero-downtime upgrade

bk201 opened the pull request:
## Support upgrading a Harvester cluster

### Start an upgrade
When backend starts, it periodically gets available versions from a remote server and records upgradable ones with `Version` resources:
```
apiVersion: harvesterhci.io/v1beta1
kind: Version
metadata:
  name: 1.0.2
  namespace: harvester-system
spec:
  isoChecksum: xxx
  isoURL: http://10.10.0.1/harvester/harvester.iso
  minUpgradableVersion: 0.3.0
  releaseDate: "20211231"
  tags:
  - dev
  - test
```

To start an upgrade, we can create an `Upgrade` resource:
```
apiVersion: harvesterhci.io/v1beta1
kind: Upgrade
metadata:
  generateName: hvst-upgrade-
  namespace: harvester-system
spec:
  version: "1.0.2"
#  image: harvester-system/image-gsjrt
```

The `version` field has to be one of the `Version` resource names. The upgrade controller then uses the ISOURL record in the version resource to start an upgrade.

For air-gapped scenarios, the user can upload an ISO to the system as a new `VirtualMachineImage`, and specify the image ID in the `image` field to start an upgrade.

### Upgrade flow

The overall flow looks like the following figure:
![upgrade-Overview - integrated drawio](https://user-images.githubusercontent.com/1691518/145429611-56cda64e-3fab-4421-852a-44698eddc9d3.png)


- **Phase1**: A VM is created from a release ISO. We call it the upgrade repo VM. The VM serves ISO contents with an Nginx server.
- **Phase2**: The upgrade controller creates a SUC plan to run jobs on each node. The job preload container images from the upgrade repo VM. The jobs use image `rancher/harvester-upgrade:<current-version>`.
- **Phase3**: A job upgrade Rancher server and all Harvester ManagedCharts. The job uses image `rancher/harvester-upgrade:<new-version>`. Because all new images are already preloaded, we can safely use the new image. The benefit is we can add new fixes here.
- **Phase4**: Upgrade the underlying RKE2 cluster. When rancher upgrades RKE2 runtime, it supports[ pre-drain and post-drain hooks](https://github.com/rancher/rancher/issues/35464). We integrate the nodes upgrade into the process:
     - Pre-drain: VMs on a node are live-migrated out. Non-migratable VMs will be shut down. 
     - Drain a node
     - Post-drain: Patch the OS image and reboot.
     - After the node is back, continue to the next node.
- **Phase5**: Clean up. The repo VM is deleted after the upgrade is deleted.

Note for single-node cases, **Phase 4** works differently. Because there is only one node, thus we can't rely on the pre-drain/post-drain mechanism. The controller creates a job to do the node upgrade:
- Download needed file from repo VM.
- Shutdown All VMs.
- Upgrade RKE2.
- Patch OS and reboot.

### CRD change

A new field is added to VirtualMachineImage CRD:
```
type VirtualMachineImageSpec struct {

<...>
	// +optional
	Checksum string `json:"checksum"`
```

The field can contain an optional checksum value (in SHA-512) to be passed to Longhorn backing image service.

### More to enhance
- [ ] Images Clean-up on hosts
- [ ] Add more tests

**Related Issue:**
https://github.com/harvester/harvester/issues/1022

**Test plan (for reviewer):**
<!-- Make sure tests pass on the Circle CI. -->
- Run a cluster with master build
- Pause `harvester-crd` ManagedChart. Otherwise, CRDs will be reverted back by Helm.
   ```
   # kubectl edit managedchart harvester-crd -n fleet-local

   spec:
      paused: true
   ```
- Delete some changed CRDs
   ```
   kubectl delete crd upgrades.harvesterhci.io
   kubectl delete crd virtualmachineimages.harvesterhci.io 
   ```
- Build a harvester image with this PR change and patch `harvester` ManagedChart to use this image. Both backend and webhook.

   e.g. (Please change the tag to fit your environment)
   ```
    $ cat patch-managedchart.json
    {
      "spec": {
        "values": {
          "replicas": 1,
          "containers": {
            "apiserver": {
              "image": {
                "imagePullPolicy": "Always",
                "repository": "rancher/harvester",
                "tag": "kiefer-dev"
              },
              "debug": true
            }
          },
          "webhook": {
            "replicas": 1,
            "image": {
              "imagePullPolicy": "Always",
              "repository": "rancher/harvester-webhook",
              "tag": "kiefer-dev"
            }
          }
        }
      }
    }

    $ cat patch-managedchart.sh
    #!/bin/bash
    kubectl patch managedcharts.management.cattle.io harvester --namespace fleet-local --patch-file patch-managedchart.json  --type merge   
   ```
- Build upgrade images and push them to somewhere for use.
   ```
   make package-upgrade
   ```

   Push the image. Note this part is tricky. During the upgrade, two versions of this image are used.
   - `harvester-upgrade:<current_version>`:  `kubectl get settings.harvesterhci.io  server-version`
   - `harvester-upgrade:<next_version>`: see next step
- Build an ISO from harvester-installer master code. Create a dump commit to make version changes. Store the ISO in a HTTP server. The ISO contains a file `harvester-release.yaml`, get the `harvester: ` version from there to use as the `<next_version>` tag in the previous step.
- Create versions CR, this pretends there are new upgrades:
   ```
   $ cat version.yaml
   apiVersion: harvesterhci.io/v1beta1
   kind: Version
   metadata:
     name: 1.0.2
     namespace: harvester-system
   spec:
     isoChecksum: d8f5c1fb41f8a6b7b8d2913223cc1da492279f1194b63d628046debf75eb647849b4fbf485725e6c4f4435fd70d6dfa9a9c0bf4c1e7611df56b37f0819aedd9a                               <-- the build ISO checksum
     isoURL: http://10.10.0.1/harvester/harvester.iso      <-- the build ISO
     minUpgradableVersion: 0.3.0
     releaseDate: "20211231"
     tags:
     - dev                 <--- `dev` tag is important
     - test

    $ kubectl create -f version.yaml
     ```
- Create an upgrade:
   ```
   $ cat upgrade.yaml
   apiVersion: harvesterhci.io/v1beta1
   kind: Upgrade
   metadata:
     generateName: hvst-upgrade-
     namespace: harvester-system
   spec:
     version: "1.0.2"

   $ kubectl create -f upgrade.yaml
   ```

bk201 committed:
Support zero-down time upgrade

- Update the upgrade controller to support zero-down time upgrade.
- Verify checksum when creating image from ISO

Signed-off-by: Kiefer Chang <kiefer.chang@suse.com>

bk201 committed:
Add upgrade scripts

Signed-off-by: Kiefer Chang <kiefer.chang@suse.com>

bk201 committed:
Workaround for Longhorn Protobuf issue

See https://github.com/longhorn/longhorn/issues/3343

Signed-off-by: Kiefer Chang <kiefer.chang@suse.com>

bk201 committed:
Add Version CRD and load upgrade-able versions

Signed-off-by: Kiefer Chang <kiefer.chang@suse.com>

bk201 committed:
Update vendors

- Bump rancher.
- Add `sigs.k8s.io/cluster-api `.

Signed-off-by: Kiefer Chang <kiefer.chang@suse.com>

bk201 committed:
Generate machines.cluster.x-k8s.io clients and controller interfaces

Signed-off-by: Kiefer Chang <kiefer.chang@suse.com>

bk201 committed:
VirtualMachineImage: support checksum verification

- Add an optional field `checksum` to CRD.
- Pass the checksum value to Longhorn backing image service.

Signed-off-by: Kiefer Chang <kiefer.chang@suse.com>

bk201 committed:
Publish upgrade image in Drone CI

Signed-off-by: Kiefer Chang <kiefer.chang@suse.com>

bk201 committed:
Make version field optional

The field is not required in air-gapped upgrade case.

Signed-off-by: Kiefer Chang <kiefer.chang@suse.com>

bk201 committed:
Workaround of undefined variable in `sigs.k8s.io/cluster-api/api/v1alpha4` package

Patch the undefined variable when generating codes.

Signed-off-by: Kiefer Chang <kiefer.chang@suse.com>

weihanglo reviewed (commented):
Looks good to me overall. Just need some time to test it manually 😆

bk201 committed:
Update chart values to include upgrade image

harvester-installer extracts the image list from values file of harvester chart.

Signed-off-by: Kiefer Chang <kiefer.chang@suse.com>

gitlawr reviewed (approved):
LGTM

guangbochen reviewed (approved):
LGTM, @bk201 can we squash some of the commit, thanks
